### ISSUE TITLE
Add a `$private` property for tokens

### ISSUE THREAD

**Issue by c1rrus** [design-tokens/community-group#110]
Should our spec add an optional `$private` property, along similar lines to [the one proposed for StyleDictionary](https://github.com/amzn/style-dictionary/pull/704) by @silversonicaxel? It would instruct tools not to display or export a token by default (though tools may provide an option for users to override that and still see / export private tokens if they wish). Its value is a boolean. `true` makes the token private, `false` makes it (explicitly) public. If there is no `$private` property, the token is public.

As per the thread on the StyleDictionary issue, I think this could be convenient way for teams to exclude (or flag) tokens that they don't consider part of their public API from code, design tools, styelguides, etc. 

For example, the following DTCG token file:

```jsonc
{
  "red": {
    "$type": "color",
    "$value": "#ff0000"
  },
  "green": {
    "$type": "color",
    "$value": "#00ff00",
    "$private": true
  },
  "blue": {
    "$type": "color",
    "$value": "#0000ff",
    "$private": false
  }
}
```

...might cause a SASS export tool to output code like this:

```scss
$red: #ff0000;
$blue: #0000ff;
```
(Note how the "green" token is omitted because it is private)

Furthermore, if a token is an alias to a private token, then the dereferenced value must be output. E.g.:

```json
{
  "red": {
    "$type": "color",
    "$value": "#ff0000"
  },
  "blue": {
    "$type": "color",
    "$value": "#00ff00",
    "$private": true
  },

  "danger-color": {
    "$value": "{red}"
  },
  "link-color": {
    "$value": "{blue}"
  }
}
```

...might be exported to SASS like so:

```
$red: #ff0000;
$danger-color: $red;
$link-color: #0000ff;
```
(Note how `$link-color` has the dereferenced value, because the "blue" token is private)

Finally, I'd suggest that this property should be inheritable from groups, just like `$type` is. Thiis would make it easier to make all tokens within a group private. For example:

```json
{
  "color": {
    "$type": "color",
    "$private": true,

    "black": {
      "$value": "#000000"
    },
    "white": {
      "$value": "#ffffff"
    }
  }
}
```

is equivalent to:

```json
{
  "color": {
    "$type": "color",

    "black": {
      "$value": "#000000",
      "$private": true
    },
    "white": {
      "$value": "#ffffff",
      "$private": true
    }
  }
}
```

What do you think?

**Comment by jeromefarnum** [comment]
I really like the idea of of the `$private` property. We have, and use, a similar property in our token system. In addition to the use cases listed above, it’s also helpful to exclude color and dimensional operation tokens that are handy shortcuts for replicating modifications throughout our token themes that we don’t want to expose through the published tokens.

**Comment by silversonicaxel** [comment]
Just a note regarding this issue

Here I already setup a PR to achieve pretty much 
https://github.com/amzn/style-dictionary/pull/704
But the idea was abandoned, since it is a breaking change, and Style Dictionary people were trying to avoid this scenario.

That's why it was suggested and implemented the idea of built-in filters.

You can raise the discussion there. 

I'm fine with the way I implemented, since I had scenarios where few design tokens needed to be filtered out. If you need to remove a group, it sounds logical wrong to have an entire set of design tokens that needed to be removed, but maybe I'm not foreseeing a situation like this.

In that case, a custom filter could be a solution?

**Comment by TravisSpomer** [comment]
I like this idea, but I wonder if `true`/`false` isn't sufficient. I can see two different "private" token possibilities:

* This token is a "base unit" of the design system but we don't need it exported to code (what we've been talking about)
* This token should never be used by designers, but we do want it exported to code

What would the latter look like? The [Salesforce design system](https://www.lightningdesignsystem.com/design-tokens/) specifies a token `brand.background.dark` which is a pretty normal color token, but also `brand.background.dark.transparent` which is always defined to be the same color with an alpha value of 0. No one should ever be changing that token, but engineers would still want that value to be available in code. There might be examples of tokens that are perfectly valid but for whatever reason the design system maintainers don't want them showing up in design tools like Figma.

`"$private": true` isn't clear enough to say which of those two cases we're talking about. I like the words `$hidden` and `$export` to specify tokens to hide from design tools and tokens to be filtered out of exported code respectively, but I don't like how one is a "positive" word and one's a "negative" word; maybe the former could be `$visible` instead. So for the previous cases in this thread, instead of marking `"$private": true` you'd add `"$export": false`, and for something like the Salesforce transparent brand color you'd use `"$visible": false`.

Or maybe that idea of "for engineers only" tokens is too small of an edge case to matter?

**Comment by chazzmoney** [comment]
@TravisSpomer This is an interesting point. I'm not sure if the specific use case is important - your point that the term "private" is ambiguous is spot on. We have a goal here (to avoid export) and a term that we are considering using to define that ("private"), and maybe that could make things confusing in the future as it isn't explicit enough. If someone else has a different mental model of what "private" means, it could be problematic.

**Comment by MatthiasDunker** [comment]
Some other wording ideas: `expose: false`, `hide: true`, `helper: true`.

**Comment by CITguy** [comment]
To piggyback off of the "private" vs "internal" discussion...

- **private**: exist after parsing
- **internal**: do not exist after parsing

## Private
A private token is one that is still included in code after transformation has taken place (see my [previous comment](https://github.com/design-tokens/community-group/issues/110#issuecomment-1148941702)).  The responsibility lies with the translation tool to identify "private" members, regardless of how private content is defined by the spec (whether it be `_` convention or explicit configuration).

## Internal
An internal token is one that is not explicitly defined in code _after_ transformation has taken place.  

If the spec allowed it, I'd be inclined to define internals using `$defs` (similar to how SVG `<defs>` behaves) and apply them using `$ref` (similar to SVG `<use>`).

```javascript
{
  // $defs will NOT be used for token _discovery_, but
  // will be used for token property _resolution_
  "$defs": {
    "colors": {
      "$type": "color",
      "green": {
        "$value": "#00ff00"
      }
    }
  },
  "red": {
    "$type": "color",
    "$value": "#ff0000"
  },
  "blue": {
    "$type": "color",
    "$value": "#0000ff"
  },
  "grass": {
    // alias of an "internal" token
    "$ref": "#/$defs/colors.green"
  }
}
```
This would resolve to the following flattened hierarchy...
```json
```

### Internal Examples
If `_green` were an "internal" token in my previous comment, you'd never see any reference to it in generated code, you'd only have the resolved value.

_example (JavaScript: ESM format)_
```javascript
export const red = '#ff0000';
export const blue = '#0000ff';
export const grass = '#00ff00';
```

_example (CSS)_
```css
:root {
  --red: #ff0000;
  --blue: #0000ff;
  --grass: #00ff00;
}
```

**Comment by Sidnioulz** [comment]
My org is also in a situation where we want more granularity than private or public. We have:
* Primary tokens that are exposed first in our documentation, autocompletion, etc and exported to code platforms
* Secondary tokens for edge cases (esp. state management) that are hidden by default in the doc to help newcomers focus on the core tokens
* Occasionally, tokens we create only as fixes to rare edge cases _within the DS code library_ (and intended to be temporary)
* Internal tokens that are used as value refs (e.g. raw color tokens) and we don't want to expose to consumers or even define in exported files

I'm not a fan of wording like `private` because it's got very specific semantics in OOP, that differ from the lay understanding of designers. A designer may expect a private token is used inside the DS component library but not exposed to consumers. Another designer may expect it to not be exported. If we map private tokens to a `private` property in Swift color classes (the OOP meaning), neither mental model is correct.

Besides, whether or not a (raw) token is exported is very easy to manage at the moment. One only has to filter their tokens prior to exporting them in Style Dictionary or similar tools. So building a rigid semantic into the standard just to spare ourselves a one-liner filter feels like over-engineering.

So I'm not sure we would benefit from a semantic enforced directly in the standard unless it is extensible or flexible. One way this could be achieved is, instead of using words with a fixed meaning, allowing a `visibility` property with a numerical value, so we can define orders of visibility. All tool makers can easily handle order relations defined by numbers, without having to attribute meaning. And tools like SD or Cobalt can define default semantics that can easily be overridden.

An order relationship would support many cases and allow extending them, e.g.:
* 1: public tokens
* 2: private or internal tokens

Or:
* 1: primary/core public tokens
* 2: secondary/extra public tokens
* 3: DS-only exported tokens
* 4: tokens exported to code but not usable directly (their name's only visible in debugging tools as they are used as value refs)
* 5: tokens that aren't exported at all (their value is always resolved when they're referenced)
* 6: forbidden tokens whose value must not be referenced at all, which might help with error reporting when deprecating tokens in a large distributed system

And anyone would be free to add however many extra layers of visibility they need and to configure their export tools to map the order relation onto concepts supported by the tools.

**Comment by nesquarx** [comment]
I believe a simple $meta array variable which can classify tokens for
categorizing, processing and extraction purposes would suffice any needs.

For example
$meta: ["private", "internal", "debug", "lightThemeOnly"];

Or whatever combination, and you can call upon tokens using another
combination of meta keywords, lets the processing define its own paradigms
than sticking to specific cases.

**Comment by Sidnioulz** [comment]
@nesquarx at the end of the day, yes we'll have access to arbitrary attributes with no meaning attached. I'm ok with going down that path and handling the meaning-making of `$private` myself, but I believe this ticket is about attributing meaning _in the spec_ :)

**Comment by nesquarx** [comment]
Agreed, if the most common cases call for $private, having it makes sense,
but trying to increase the scope there instead of an additional generic
solution might become exhausting.
